src/compiler/checker.ts

```diff
@@ -21000,9 +21000,14 @@ namespace ts {
          * with no call or construct signatures.
          */
         function isObjectTypeWithInferableIndex(type: Type): boolean {
-            return type.flags & TypeFlags.Intersection ? every((type as IntersectionType).types, isObjectTypeWithInferableIndex) :
-                !!(type.symbol && (type.symbol.flags & (SymbolFlags.ObjectLiteral | SymbolFlags.TypeLiteral | SymbolFlags.Enum | SymbolFlags.ValueModule)) !== 0 &&
-                !typeHasCallOrConstructSignatures(type)) || !!(getObjectFlags(type) & ObjectFlags.ReverseMapped && isObjectTypeWithInferableIndex((type as ReverseMappedType).source));
+            return type.flags & TypeFlags.Intersection
+                ? every((type as IntersectionType).types, isObjectTypeWithInferableIndex)
+                : !!(
+                    type.symbol
+                    && (type.symbol.flags & (SymbolFlags.ObjectLiteral | SymbolFlags.TypeLiteral | SymbolFlags.Enum | SymbolFlags.ValueModule)) !== 0
+                    && !((type.symbol.flags & SymbolFlags.Class) && (type.symbol.flags & SymbolFlags.ValueModule))

```

Comment:
Is there a reason not to just use `!(type.symbol.flags & SymbolFlags.Class)` (dropping the “and is also a value module”)? It looks to me like we believe namespaces generally have an inferable index, and classes don’t, and we want the classiness property to trump the namespacey property in this calculation. It doesn’t seem to need the specificity of “classes merged with namespaces.”

Reply:
Probably not; things still work as expected with that simplification. I just took the solution suggested in the issue and thumbs-up'd by Ryan.

Reply:
Switched it over.